test/hotspot/jtreg/compiler/lib/template_framework/StructuralName.java

```diff
@@ -0,0 +1,197 @@
+ *               {@link StructuralName} when sampling later on.
+ */
+public record StructuralName(String name, StructuralName.Type type, int weight) implements Name {
+
+    /**
+     * Creates a new {@link StructuralName}.
+     */
+    public StructuralName {
+    }
+
+    /**
+     * The interface for the type of a {@link StructuralName}.
+     */
+    public interface Type extends Name.Type {
+        /**
+         * The name of the type, that can be used in code.
+         *
+         * @return The {@link String} representation of the type, that can be used in code.
+         */
+        String name();
+
+        /**
+         * Defines the subtype relationship with other types, which is used to filter {@link StructuralName}s
+         * in {@link View#exactOf}, {@link View#subtypeOf}, and {@link View#supertypeOf}.
+         *
+         * @param other The other type, where we check if it is the supertype of {@code 'this'}.
+         * @return If {@code 'this'} is a subtype of {@code 'other'}.
+         */
+        boolean isSubtypeOf(StructuralName.Type other);
+    }
+
+    /**
+     * The {@link View} represents a filtered set of {@link StructuralName}s in the current scope.
+     * It can be obtained with {@link Template#structuralNames}. It can be used to count the
+     * available {@link StructuralName}s, or sample a random {@link StructuralName} according to the
+     * weights of the {@link StructuralName}s in the filtered set.
+     */
+    public static final class View {
+        private final StructuralName.Type subtype;
+        private final StructuralName.Type supertype;
+
+        View(StructuralName.Type subtype, StructuralName.Type supertype) {
+            this.subtype = subtype;
+            this.supertype = supertype;
+        }
+
+        View() {
+            this(null, null);
+        }
+
+        NameSet.Predicate predicate() {
+            if (subtype == null && supertype == null) {
+                throw new UnsupportedOperationException("Must first call 'subtypeOf', 'supertypeOf', or 'exactOf'.");
+            }
+            return (Name name) -> {
+                if (!(name instanceof StructuralName dn)) { return false; }
+                if (subtype != null && !dn.type().isSubtypeOf(subtype)) { return false; }
+                if (supertype != null && !supertype.isSubtypeOf(dn.type())) { return false; }
+                return true;
+            };
+        }
+
+        /**
+         * Create a filtered {@link View}, where all {@link StructuralName}s must be subtypes of {@code type}.
+         *
+         * @param type The type of which all {@link StructuralName}s must be subtypes of.
+         * @return The filtered {@link View}.
+         * @throws UnsupportedOperationException If this {@link View} was already filtered with
+         *                                       {@link subtypeOf} or {@link exactOf}.
```

Comment:
Same here and in methods below:
```suggestion
         *                                       {@link #subtypeOf} or {@link #exactOf}.
```